src/functional/iterator/transforms/inline_lambdas.py

```diff
@@ -1,32 +1,94 @@
 import dataclasses
+from typing import Optional
 
-from eve import NodeTranslator, NodeVisitor
+from eve import NodeTranslator
 from functional.iterator import ir
+from functional.iterator.transforms.count_symbol_refs import CountSymbolRefs
 from functional.iterator.transforms.remap_symbols import RemapSymbolRefs, RenameSymbols
 
 
-@dataclasses.dataclass
-class CountSymbolRefs(NodeVisitor):
-    ref_counts: dict[str, int]
-
-    @classmethod
-    def apply(cls, node: ir.Node, symbol_names: list[str]) -> dict[str, int]:
-        ref_counts = {name: 0 for name in symbol_names}
-        active_refs = set(symbol_names)
-
-        obj = cls(ref_counts=ref_counts)
-        obj.visit(node, active_refs=active_refs)
-
-        return ref_counts
-
-    def visit_SymRef(self, node: ir.Node, *, active_refs: set[str]):
-        if node.id in active_refs:
-            self.ref_counts[node.id] += 1
-
-    def visit_Lambda(self, node: ir.Lambda, *, active_refs: set[str]):
-        active_refs = active_refs - set(param.id for param in node.params)
+def inline_lambda(
+    node: ir.FunCall,
+    opcount_preserving=False,
+    force_inline_lift=False,
+    eligible_params: Optional[list[bool]] = None,
+):
+    eligible_params = eligible_params or [True] * len(node.fun.params)
+
+    assert len(eligible_params) == len(node.fun.params) == len(node.args)
+
+    if opcount_preserving:
+        ref_counts = CountSymbolRefs.apply(node.fun.expr, [p.id for p in node.fun.params])
+
+        for i, param in enumerate(node.fun.params):
+            # TODO(tehrengruber): allow inlining more complicated zero-op expressions like
+            #  ignore_shift(...)(it_sym)  # noqa: E800
+            if ref_counts[param.id] != 1 and not isinstance(
+                node.args[i], (ir.SymRef, ir.Literal, ir.OffsetLiteral)
+            ):
+                eligible_params[i] = False
+
+    if force_inline_lift:
+        for i, arg in enumerate(node.args):
+            if (
+                isinstance(arg, ir.FunCall)
+                and isinstance(arg.fun, ir.FunCall)
+                and isinstance(arg.fun.fun, ir.SymRef)
+                and arg.fun.fun.id == "lift"
+            ):
+                eligible_params[i] = True
+
+    if len(eligible_params) != 0 and not any(eligible_params):
```

Comment:
Maybe a bit more clear what’s checked:
```suggestion
    if node.fun.params and not any(eligible_params):
```